### DISCUSSION TITLE
Try to save the transformed model to mps/lp

### DISCUSSION THREAD

**Discussion by zhejiangzhoulei** [ds4dm/ml4co-competition#38] [Q&A] [answered]
Hi all,
I'm trying to save the transformed MIP model to a file.
Take Instance load_balancing_9900 for example.
When I save the transformed model to a mps/lp file
```
    def extract(self, model, done):
        if done:
            return None

        m = model.as_pyscipopt()
        variables = m.getVars(transformed=True)
        inf = m.infinity()

        # extract the upper and lower bounds of all variables
        lbs = np.asarray([v.getLbLocal() for v in variables])
        ubs = np.asarray([v.getUbLocal() for v in variables])
        has_lb = np.asarray([lb > -inf for lb in lbs])
        has_ub = np.asarray([ub < inf for ub in ubs])

        observation = (m,has_lb, has_ub, lbs, ubs)

        if self.needwrite:
            m.writeProblem(filename='tranformed_problem.mps', trans=True, genericnames=False)
            m.writeProblem(filename='tranformed_problem.lp', trans=True, genericnames=False)
            self.needwrite=False
        print("len(var):"+str(len(variables)))

        return observation
```
the transformed model has 3987 variables
```
Instance load_balancing_9900.mps.gz
  seed: 0
  initial primal bound: 889.0
  initial dual bound: 657.89004724696
  objective offset: 0
wrote problem to file /home/*******/example/tranformed_problem.mps
wrote problem to file /home/*******/example/tranformed_problem.lp
len(var):3987
```
However, when I use scip to read it
<img width="680" alt="屏幕快照 2021-07-26 下午4 30 51" src="https://user-images.githubusercontent.com/10865488/126958591-9e3084bc-0527-409b-b6f9-9d1a3a42d7dd.png">

I do not know why the model saved in the mps file has so many variables

Any help or advice would be greatly appreciated.
Best

**Comment by ambros-gleixner** [comment, marked as answer]
This is indeed strange, can you post the two files tranformed_problem.mps and lp?

**Reply by zhejiangzhoulei** [reply]
hi, @ambros-gleixner, here are the two files.
[files.zip](https://github.com/ds4dm/ml4co-competition/files/6879033/files.zip)

**Reply by ambros-gleixner** [reply]
I see, the MPS writer also writes the variables that were fixed during presolving. You can see that in the lines with "FX Bound". I don't see a way to deactivate this by parameter easily, but it's not really a problem. You could remove these sections in the file, and also SCIP will remove them again in the first presolving round.

**Reply by zhejiangzhoulei** [reply]
@ambros-gleixner , thank you :)

**Reply by th-yoon** [reply]
Hi @ambros-gleixner, I'm trying to read .lp file saved and it emits parsing error at SCIP reader C code. Do you have any idea how to handle this issue? Please find the attached error message and the corresponding lp file.
![image](https://user-images.githubusercontent.com/54887560/129579042-25474a8a-ebff-4ba3-b951-78f151141ec3.png)
[anonymous_14.zip](https://github.com/ds4dm/ml4co-competition/files/6993067/anonymous_14.zip)

**Reply by ambros-gleixner** [reply]
Can it be that the line above is the problem?

**Reply by th-yoon** [reply]
Sure it seems. I wonder if this parsing error is intended.

**Reply by ambros-gleixner** [reply]
No, I checked, actually the name of the constraint is the problem: "4298_lhs" is interpreted as 4298 times a variable called "_lhs". You can fix it by making sure constraint names always start with letters, not numbers.
